# The main entry point of your workflow.
# After configuring, running snakemake -n in a clone of this repository should successfully execute a dry-run of the workflow.
import itertools
import os 
import glob
import sys
import pandas as pd

configfile: "../config/config.yaml"

#report: "report/workflow.rst"
include: "rules/common.smk"
include: "rules/metagenomics.smk"
localrules: multiqc_trimmed

# Allow users to fix the underlying OS via singularity.
rule all:
    input:
        expand("../results/filtered/{dataset}/{sample}.filtered.R1.fastq.gz",
            sample=samples["sample"], dataset=samples["dataset"]),
        expand("../results/filtered/{dataset}/{sample}.filtered.R2.fastq.gz", 
            sample=samples["sample"], dataset=samples["dataset"])
        #expand("../results/filtered/{dataset}/{sample}.filtered.R2.fastq.gz")
