
# 20230131
Processing LyPMA fragmentation optimization

Changed alignment genome back to human from mouse

Updated prep_sample_sheet.py script so it sorts sample sheet alphabetically (easier navigation)

Command line code for making sample sheet   
    python3 workflow/scripts/prep_sample_sheet.py --delimiter="_S" --path="/projects/p31648/bmoe008_LibraryFragmentLength/" --subdirectories=True

Would like to clean up bowtie2 mapping and implement a compression step on fastq files

Worked! Got much cleaner version of bowtie2 that cuts out all intermediate bam/sam files except final, mapped, sorted bam
    also unmapped reads now in fastq.gz

20230201

AHHHHHHHHHH WHYYYYYYYY
cluster deleted all conda envs and older mlm files that haven't been modified in awhile
forced to start a new mlm repo, fastest solution

AHHHHH -- snakemamba directory now isnt working, have to reinstall too...
deleted and tried reinstalling from yaml, but snakemake has error like reported in 20221103_notes.md

solution
    make sure to check ~/.condarc
        apparently conda-forge doesnt play nice with defaults channel anymore(?)/with mamba
        deleted biocore, defaults, and biobakery from channels list
	added back defaults

ok so this solution took hours to get around and, not gonna lie, was really rough
- deleted all conda environments (i.e., rm -rf .conda)
- from now on, we're fully switching to mamba module
- did mamba init bash and then restarted shell
- upon relogin, has "(base)" active since mamba was initiated
- mamba create -c conda-forge -c bioconda -n snakemake snakemake
- started working again

**Need** to containerize conda envs to prevent this from happening agains...

also had to manually "conda update fastp" to change fastp from 0.20.1 > 0.22.0
	changes in the conda yaml in case 
	older version, which somehow was being updated in other mlm, doesn't support dedup


