---
title: "General ReadQC and AssemblyQC analysis"
output: html_notebook
---

```{r}
library(tidyverse)
library(ggplot2)
library(ggpubr)
library(cowplot)
library(ggsci)
```

```{r}
# Set multiqc file name and summary output name
readqc = "../../results/fastqc_out/multiqc_data/multiqc_fastqc.txt"
summary_out = "../../results/tables/ReadNumberSummary.tsv"

# Parse multiqc/fastqc data to have columns for Sample and Tool+Read# 
#   (e.g., bowtie_qc.1, bowtie_qc.2, bbmerge.merged)
tmp_readqc<- read_tsv(readqc, name_repair = "universal") %>% 
  select(Sample,Total.Sequences,Sequence.length,.GC, 
         total_deduplicated_percentage, avg_sequence_length) %>%
  separate(Sample, into = c("Tool", "Sample", "Read"), sep = "\\s*\\|\\s*") %>%
  mutate(Read = ifelse(grepl("_R1_", Read), "1",
                       ifelse(grepl("_R2_", Read), "2", 
                              ifelse(grepl("_R1", Read), "1",
                              ifelse(grepl("_R2", Read), "2",
                              ifelse(grepl(".r1", Read), "1",
                              ifelse(grepl(".r2", Read), "2", "merged")))))), .after="Read") %>%
  mutate(Tool_Read = paste(Tool, Read, sep = ".")) #%>% select(!c(Tool, Read)) 

# Make major summary/output table for total sequences by tools per 
fastqc_summary <- tmp_readqc %>% 
  select(Sample, Tool_Read, Total.Sequences) %>% 
  pivot_wider(names_from = Tool_Read, values_from = Total.Sequences) %>%
  select(Sample, raw_qc.1, raw_qc.2, fastp_qc.merged, bbduk_qc.1, bbduk_qc.2, bowtie_qc.1, bowtie_qc.2) %>% #bbmerge_qc.merged
    # mutate("fastp-raw"= fastp_qc.merged/raw_qc.1) %>% 
    # mutate("bbduk-fastp"= bbduk_qc.1/fastp_qc.merged) %>% 
    # mutate("bowtie-bbduk"= bowtie_qc.1/bbduk_qc.1) #%>% mutate("bbmerge-bowtie"= bbmerge_qc.merged/bowtie_qc.1)    
    mutate("raw.FC"= raw_qc.1/raw_qc.1) %>% 
    mutate("fastp.FC"= fastp_qc.merged/raw_qc.1) %>% 
    mutate("bbduk.FC"= bbduk_qc.1/raw_qc.1) %>% 
    mutate("bowtie.FC"= bowtie_qc.1/raw_qc.1) #%>% mutate("bbmerge-bowtie"= bbmerge_qc.merged/bowtie_qc.1)

write_tsv(fastqc_summary, summary_out)
```

```{r}
# Make major summary/output table for total sequences by tools per 
fastqc_tidy <- fastqc_summary %>% 
  select(Sample, raw.FC,fastp.FC, bbduk.FC, bowtie.FC) %>%
  pivot_longer(2:last_col()) %>%
  mutate(name = factor(name, levels=c("raw.FC","fastp.FC", "bbduk.FC", "bowtie.FC")))

fastqc_tidy %>%
  group_by(name) %>%
  summarize(min = min(value),
            q1 = quantile(value, 0.25),
            median = median(value),
            mean = mean(value),
            q3 = quantile(value, 0.75),
            max = max(value))

ggplot(fastqc_tidy, aes(x=name, y=value)) +
  geom_boxplot(aes(fill=name), outlier.alpha = 0) +
  geom_point(aes(fill=name), 
             alpha = 0.75,
             size = 1, 
             shape = 21, 
             color = 'black',
             position = position_jitterdodge(jitter.width = 1)) +
  scale_fill_simpsons() +
  nature_theme("Read", "Fol") +
  labs(x="ReadQC Step", 
       y="Read Fold Change", 
       title="Stepwise Loss of Sequencing Data") +
  theme(aspect.ratio = 1)
```

# Quast

```{r}
quast_tidy <- read_tsv("../../results/quast_out/multiqc_data/multiqc_quast.txt") %>%
  pivot_longer(2:last_col()) %>%
  separate(Sample, into=c(NA, NA, "Sample"), sep="\\| ") %>%
  left_join(., s_meta, by=c("Sample" = "seq_id")) %>%
  filter(!stringr::str_detect(name, ">")) %>%
  #mutate(value = sqrt(value)) %>%
  mutate(RepeatID=if_else(str_detect(Sample.y, "BAL00"), "Baseline", if_else(str_detect(Sample.y, "BAL"), "Repeat", NA)))
  
quast_tidy %>%
  group_by(cluster_num, name) %>%
  summarize(min = min(value),
            q1 = quantile(value, 0.25),
            median = median(value),
            mean = mean(value),
            q3 = quantile(value, 0.75),
            max = max(value))

ggplot(quast_tidy, aes(x=cluster_num, y=value)) +
  geom_boxplot(aes(fill=cluster_num), outlier.alpha = 0, alpha=0.75) +
  geom_point(aes(fill=cluster_num), 
             alpha = 0.9,
             size = 1, 
             shape = 21, 
             color = 'black',
             position = position_jitterdodge(jitter.width = .5)) +
  scale_fill_npg() +
  nature_theme("Clust", "Stat") +
  labs(x="Pneumotype Cluster", 
       y="Square Root of Quast Statistic", 
       title="Number of Contigs") +
  facet_wrap(~ name, scales = "free") +
  theme(aspect.ratio = 1, 
        strip.background = element_blank(), 
        strip.text.x = element_text(size = 5, face="bold")) 
```

```{r}
quast_tidy %>% 
  filter(cluster_num=="5") %>%
  arrange(desc(value))

```

