---
title: "02A_MetaPhlanTaxonomicAnalysis"
output: html_notebook
---

```{r}
library(tidyverse)
library(ggplot2)
library(ggpubr)
library(cowplot)
library(vegan)

```

# Read in and Prep-process data
## Read in Taxa Tables

```{r}
taxa_table = read_table('../../results/kraken/merged_kraken_mpa_profile.tsv') %>%
                        rename(clade_name = "#Classification")

# Remove _kraken2report.txt from ends of files 
colnames(taxa_table) <- sapply(strsplit(names(taxa_table), "_kraken2report.txt"), `[[`, 1)

# Remove Taxa with no samples above count threshold
min_counts <- 5
parsed_taxa <- taxa_table %>% 
  melt() %>% 
  filter(value > min_counts) %>% # filter 0 values to get prevalence
  select("clade_name") %>% 
  unique() %>% 
  pluck(1)
taxa_table <- taxa_table %>% filter(clade_name %in% parsed_taxa)
```


```{r}
genus_table <- taxa_table %>% filter(grepl("g__", clade_name), !grepl("s__", clade_name))
genus_table <- genus_table %>% column_to_rownames(var = c("clade_name"))

species_table <- taxa_table %>% filter(grepl("s__", clade_name))
species_table <- species_table %>% column_to_rownames(var = c("clade_name"))
```

## Read in mapping table
```{r}
map_table <- read_table('../../config/map_table.tsv') %>%
                        column_to_rownames(var = c("sample"))
```


## remove neg controls; set sample.type factor
```{r}
species_table <- species_table[,rownames(map_table)] # reorder to match metadata
species_table <- species_table[,-c(grep("Control", map_table$sample.type))] # remove control samples
#species_table <- species_table %>% prop.table() # Normalize to 100%

genus_table <- genus_table[,rownames(map_table)]
genus_table <- genus_table[,-c(grep("Control", map_table$sample.type))]
#genus_table <- genus_table %>% prop.table() # Normalize to 100%

map_table <- map_table[-c(grep("Control", map_table$sample.type)), ]
map_table$sample.type = factor(map_table$sample.type, levels=c("Pneumonia", "NonPneumonia", "Unknown"))

all(row.names(map_table) %in% colnames(species_table))

```

## optional, remove follow up BAL
```{r}
 species_table <- species_table[,rownames(map_table)] # rereorder to match metadata

 species_table <- species_table[,-c(grep("FALSE", map_table$is.baseline))]

 map_table <- map_table[-c(grep("FALSE", map_table$is.baseline)), ]

# set sample type as factor

 all(row.names(map_table) %in% colnames(species_table)) # Sanity check

```


####################################
# Alpha diversity by sample type 
####################################

## compute shannon and simpson diversity metrics
```{r}
diversity_vec = matrix(nrow = dim(species_table)[2], ncol = 2)
diversity_vec = as.data.frame(diversity_vec)
for (a in 1:dim(species_table)[2]) {
  diversity_vec[a,1] = diversity(species_table[,a], index = "shannon")
  diversity_vec[a,2] = diversity(species_table[,a], index = "simpson")
}
colnames(diversity_vec) = c("Shannon", "Simpson")

# add sample.type factor
diversity_vec$sample.type = map_table$sample.type
diversity_vec$sample.type = factor(diversity_vec$sample.type) # Optional: Add Levels
```


## boxplots 
```{r}
ggplot(diversity_vec, aes(x = sample.type, y = Shannon, fill = sample.type)) +
  geom_boxplot(outlier.shape = NA, alpha = 0.7) +
  #geom_violin(alpha = 0.8) +
  geom_jitter(size = 1, width = 0.1, alpha = 0.35) +
  #scale_fill_manual(values=c("blue", "lightgreen", "yellow", "brown", "violet", "gray"),
  #                  labels = c("Toothbrush", "HMP-Oral", "HMP-Skin", "HMP-Gut", "HMP-Vaginal", "Building dust")) +
  theme_bw() +
  theme(legend.position = "none",
        axis.title.x = element_blank(),
        #axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        axis.title.y = element_text(size = 14),
        axis.text.y = element_text(size = 14))
```
## ANOVA stats
```{r}
summary(aov(Shannon ~ sample.type, diversity_vec))
TukeyHSD(aov(Shannon ~ sample.type, diversity_vec))
```
####################################
# Beta diversity between sample types 
####################################

```{r}
# beta-diversity measure
beta <- vegdist(t(species_table), 'jaccard', binary = T)
beta.mat <- as.matrix(beta)

# projection
pcoa <- cmdscale(beta, k = 4, eig = TRUE)

# cleanup
ord <- as.data.frame(pcoa$points)
names(ord) <- c('pcoa1', 'pcoa2', 'pcoa3', 'pcoa4') ## rename coordinates


# add metadata
ord$Category = map_table$sample.type
ord$Outcome = map_table$binned_outcome
ord$Baseline = map_table$is.baseline
ord$Diagnosis = map_table$diagnosis.subtype

# Percent explained variation
eig <- eigenvals(pcoa)
eig.percent <- 100*head(eig/sum(eig))
eig.percent
eig.1 <- paste("PCo1 (", as.character(round(eig.percent[1], digits = 1)), "%)", sep = "")
eig.2 <-paste("PCo2 (", as.character(round(eig.percent[2], digits = 1)), "%)", sep = "")

```

```{r}
## plot PCoA (FIGURE 1A)
ggplot(data = ord, aes(x = pcoa1, y = pcoa2, fill = Outcome)) +
  geom_point(size = 5, stroke = 1.5, shape=21) +
  theme_bw() +
  xlab(eig.1) +
  ylab(eig.2) +
  theme_q2r() +
  theme(axis.text = element_text(size=15, color = "black"),
        axis.title = element_text(size=16, color = "black"),
        legend.text = element_text(size=15, color = "black"),
        legend.title = element_text(size=15, color = "black")) +
  scale_fill_brewer(guide="legend", palette="Set1") +
  #scale_shape_manual(values=c(21))+#, 22, 23, 24, 25, 26)) +
  guides(fill=guide_legend(override.aes=list(shape=21))) +
   theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) 
ggsave("../../results/notebook_out/02A_KrakenPCoA.pdf", units="in", width=6, height=3.5)

```

## PERMANOVA Stats
```{r}
# effect of sample type
beta <- vegdist(t(species_table), 'jaccard', binary = T)
adonis_out <- adonis2(beta ~ overall, data = map_table, permutations = 999)
adonis_out
```

## Distance between samples
```{r}
# effect of sample type
beta <- vegdist(t(species_table), 'jaccard', binary = T)
adonis2(beta ~ diagnosis.subtype, data = map_table, permutations = 999)

```

## Between diagnostic types
```{r}
# Bac vs Viral
beta <- vegdist(t(species_table[,grep("Other_Pneumonia|Other_Viral_Pneumonia",map_table$diagnosis.subtype)]), 'jaccard', binary = T)
adonis2(beta ~ diagnosis.subtype, data = map_table[grep("Other_Pneumonia|Other_Viral_Pneumonia",map_table$diagnosis.subtype),], permutations = 999)
```

```{r}
# Bac vs NonP

beta <- vegdist(t(species_table[,grep("Other_Pneumonia|NonPneumonia",map_table$diagnosis.subtype)]), 'jaccard', binary = T)
adonis2(beta ~ diagnosis.subtype, data = map_table[grep("Other_Pneumonia|NonPneumonia",map_table$diagnosis.subtype),], permutations = 999)
```
```{r}
# Bac vs Unknown

beta <- vegdist(t(species_table[,grep("Other_Pneumonia|Unknown",map_table$diagnosis.subtype)]), 'jaccard', binary = T)
adonis2(beta ~ diagnosis.subtype, data = map_table[grep("Other_Pneumonia|Unknown",map_table$diagnosis.subtype),], permutations = 999)
```

```{r}
# Viral vs Non

beta <- vegdist(t(species_table[,grep("Other_Viral_Pneumonia|NonPneumonia",map_table$diagnosis.subtype)]), 'jaccard', binary = T)
adonis2(beta ~ diagnosis.subtype, data = map_table[grep("Other_Viral_Pneumonia|NonPneumonia",map_table$diagnosis.subtype),], permutations = 999)
```

```{r}
# Viral vs Unknown

beta <- vegdist(t(species_table[,grep("Other_Viral_Pneumonia|Unknown",map_table$diagnosis.subtype)]), 'jaccard', binary = T)
adonis2(beta ~ diagnosis.subtype, data = map_table[grep("Other_Viral_Pneumonia|Unknown",map_table$diagnosis.subtype),], permutations = 999)
```

### Between outcomes
```{r}
# effect of sample type
beta <- vegdist(t(species_table), 'jaccard', binary = T)
adonis2(beta ~ binned_outcome, data = map_table, permutations = 999)

```
```{r}
# home vs deceased
beta <- vegdist(t(species_table[,grep("Home|Deceased",map_table$binned_outcome)]), 'jaccard', binary = T)
adonis2(beta ~ binned_outcome, data = map_table[grep("Home|Deceased",map_table$binned_outcome),], permutations = 999)
```
```{r}
# home vs ltac
beta <- vegdist(t(species_table[,grep("Home|LTAC",map_table$binned_outcome)]), 'jaccard', binary = T)
adonis2(beta ~ binned_outcome, data = map_table[grep("Home|LTAC",map_table$binned_outcome),], permutations = 999)
```
```{r}
# home vs ltac
beta <- vegdist(t(species_table[,grep("Home|Inpatient_Facility",map_table$binned_outcome)]), 'jaccard', binary = T)
adonis2(beta ~ binned_outcome, data = map_table[grep("Home|Inpatient_Facility",map_table$binned_outcome),], permutations = 999)
```
```{r}
# Deceased vs ltac
beta <- vegdist(t(species_table[,grep("Deceased|LTAC",map_table$binned_outcome)]), 'jaccard', binary = T)
adonis2(beta ~ binned_outcome, data = map_table[grep("Deceased|LTAC",map_table$binned_outcome),], permutations = 999)
```

```{r}
# Deceased vs ltac
beta <- vegdist(t(species_table[,grep("Deceased|Inpatient",map_table$binned_outcome)]), 'jaccard', binary = T)
adonis2(beta ~ binned_outcome, data = map_table[grep("Deceased|Inpatient",map_table$binned_outcome),], permutations = 999)
```

```{r}
sample.type.1 <- map_table$sample.type[1]
sample.type.1

mean(as.matrix(vegdist(t(species_table[,grep(sample.type.1,map_table$sample.type)]), 
                       'jaccard', binary = T)))
sd(as.matrix(vegdist(t(species_table[,grep(sample.type.1,map_table$sample.type)]), 
                     'jaccard', binary = T)))/sqrt(34)
```

####################################
# Heat Map Most Abundant Taxa
####################################

## Get most abundant, prevalent species in ranked order
```{r}
min_mean_proportion <- .00001
min_prevalence <- 20

species.ranked <- species_table %>% 
  rownames_to_column() %>% 
  melt() %>% 
  filter(value > 0) %>% # filter 0 values to get prevalence
  group_by(rowname) %>%
  summarise(mean = mean(value), n = n(), sum=sum(value)) %>% 
  arrange(desc(mean)) %>% 
  #filter(mean > min_mean_proportion) %>% # Filter by mean percent
  filter(n > min_prevalence) %>% # Filter by prevalence
  pluck(1)

length(species.ranked)
```

## Order by species rank object and remove extra taxnomic info
```{r}
species_table.heat <- species_table[species.ranked,] %>% 
  rownames_to_column() %>%
  separate(rowname, into = c("ExtraTaxa", "Taxa"), sep="s__") %>%
  select(!ExtraTaxa) %>% 
  column_to_rownames("Taxa") %>% 
  as.matrix
```

```{r}
library(pheatmap)
annot.col <- map_table %>% select(sample.type, binned_outcome, is.baseline) 

paletteLength <- 100

myColors <- rev(colorRampPalette(rev(brewer.pal(n = 9, name ="Reds")))(paletteLength))
myColors <- rev(colorRampPalette(rev(brewer.pal(n = 9, name ="RdBu")))(paletteLength))

#pdf(file="../../results/notebook_out/02A_KrakenPlot.pdf", width = 25, height = 20)
pheatmap(species_table.heat, 
         color = myColors,
         annotation_col = annot.col,
         angle_col = "45",
         show_colnames=FALSE,
         show_rownames = FALSE, scale="row")
#dev.off()
```


## Genus
```{r}
min_mean_proportion <- .00001
min_prevalence <- 20

genera.ranked <- genus_table %>% 
  rownames_to_column() %>% 
  melt() %>% 
  filter(value > 0) %>% # filter 0 values to get prevalence
  group_by(rowname) %>%
  summarise(mean = mean(value), n = n(), sum=sum(value)) %>% 
  arrange(desc(mean)) %>% 
  #filter(mean > min_mean_proportion) %>% # Filter by mean percent
  filter(n > min_prevalence) %>% # Filter by prevalence
  pluck(1)

length(genera.ranked)
```
```{r}
# Order by species rank object and remove extra taxnomic info
genus_table.heat <- genus_table[genera.ranked,] %>% 
  rownames_to_column() %>%
  separate(rowname, into = c("ExtraTaxa", "Taxa"), sep="g__") %>%
  select(!ExtraTaxa) %>% 
  column_to_rownames("Taxa") %>% 
  as.matrix
```

```{r}
library(pheatmap)
annot.col <- map_table %>% select(sample.type, binned_outcome, is.baseline) 

paletteLength <- 100

myColors <- rev(colorRampPalette(rev(brewer.pal(n = 9, name ="Reds")))(paletteLength))
myColors <- rev(colorRampPalette(rev(brewer.pal(n = 9, name ="RdBu")))(paletteLength))

#pdf(file="../../results/notebook_out/02A_KrakenPlot.pdf", width = 25, height = 20)
pheatmap(genus_table.heat, 
         color = myColors,
         annotation_col = annot.col,
         angle_col = "45",
         show_colnames=FALSE,
         show_rownames = FALSE, scale="row")
#dev.off()
```


####################################
# Testing as phyloseq + Playground
####################################

```{r}
library(tidyverse); packageVersion("tidyverse")     #version:1.3.0 
library(phyloseq); packageVersion("phyloseq")       #version:1.32.0


control_samples <- read_table('../../config/map_table.tsv') %>% column_to_rownames(var = c("sample")) %>%
  filter(grepl("Control", sample.type)) %>% row.names()
map_table <- read_table('../../config/map_table.tsv')

s_abund <- read_tsv("../../results/old_results/tmp_resultsB//kraken/merged_kraken_mpa_profile.tsv") %>%
  rename(clade_name = "#Classification") %>% 
  filter(grepl("s__", clade_name))


s_tax_tab <- s_abund %>%
  dplyr::rename("taxonomy" = "clade_name") %>%
  dplyr::select(taxonomy) %>%
  dplyr::mutate(Species = sub('.*\\|', '', taxonomy),
                Species = gsub("s__", "", Species),
                spec_row = Species) %>%
  dplyr::select(-taxonomy) %>%
  tibble::column_to_rownames(var = "spec_row")

s_otu_tab <- s_abund %>%
  dplyr::rename("taxonomy" = "clade_name") %>%
  dplyr::mutate(taxonomy = sub('.*\\|', '', taxonomy),
                taxonomy = gsub("s__", "", taxonomy)) %>%
  tibble::column_to_rownames(var = "taxonomy")

names(s_otu_tab) <- gsub(names(s_otu_tab), pattern = "_kraken2report.txt", replacement = "") 

s_otu_tab <-s_otu_tab %>% select(!control_samples)

s_meta <- data.frame(seq_id = names(s_otu_tab))
s_meta <- s_meta %>%
  dplyr::mutate(sampleNames_row = seq_id) %>%
  tibble::column_to_rownames(var = "sampleNames_row")
s_meta %>% left_join(map_table, by=c("seq_id" = "sample"))
(ps_kraken_species <- phyloseq(sample_data(s_meta),
                                otu_table(s_otu_tab, taxa_are_rows = TRUE),
                                tax_table(as.matrix(s_tax_tab))))

# filter taxa ("Remove taxa not seen more than 3 times in at least 20% of the samples" )
ps_kraken_species <-filter_taxa(ps_kraken_species, function(x) sum(x > 3) > (0.1*length(x)), TRUE)
ps_kraken_species
```

```{r}
library(metagenomeSeq)
heatmapCols = colorRampPalette(brewer.pal(9, "RdBu"))(50)

pdf("../../results/notebook_out/02A_KrakenCorrPlot.pdf", width = 12, height=12)
plotCorr(obj = ps_kraken_species %>% phyloseq_to_metagenomeSeq(), n = 200, cexRow = 0.5, cexCol = 0.5,
trace = "none", dendrogram = "none", col = heatmapCols)
dev.off()


pdf("../../results/notebook_out/02A_KrakenAbundPlot.pdf")
plotMRheatmap(obj = ps_kraken_species %>% phyloseq_to_metagenomeSeq(), n = 200, cexRow = 0.25, cexCol = 0.25,
trace = "none", dendrogram = "none", col = heatmapCols)
dev.off()
```

# Genus

```{r}
library(tidyverse); packageVersion("tidyverse")     #version:1.3.0 
library(phyloseq); packageVersion("phyloseq")       #version:1.32.0


control_samples <- read_table('../../config/map_table.tsv') %>% column_to_rownames(var = c("sample")) %>%
  filter(grepl("Control", sample.type)) %>% row.names()
map_table <- read_table('../../config/map_table.tsv')

s_abund <- read_tsv("../../results/kraken/merged_kraken_mpa_profile.tsv") %>%
  rename(clade_name = "#Classification") %>% 
  filter(grepl("g__", clade_name)) %>%   filter(!grepl("s__", clade_name))



s_tax_tab <- s_abund %>%
  dplyr::rename("taxonomy" = "clade_name") %>%
  dplyr::select(taxonomy) %>%
  dplyr::mutate(Genus = sub('.*\\|', '', taxonomy),
                Genus = gsub("g__", "", Genus),
                spec_row = Genus) %>%
  dplyr::select(-taxonomy) %>%
  tibble::column_to_rownames(var = "spec_row")

s_otu_tab <- s_abund %>%
  dplyr::rename("taxonomy" = "clade_name") %>%
  dplyr::mutate(taxonomy = sub('.*\\|', '', taxonomy),
                taxonomy = gsub("g__", "", taxonomy)) %>%
  tibble::column_to_rownames(var = "taxonomy")

names(s_otu_tab) <- gsub(names(s_otu_tab), pattern = "_kraken2report.txt", replacement = "") 

s_otu_tab <-s_otu_tab %>% select(!control_samples)

s_meta <- data.frame(seq_id = names(s_otu_tab))
s_meta <- s_meta %>%
  dplyr::mutate(sampleNames_row = seq_id) %>%
  tibble::column_to_rownames(var = "sampleNames_row")
s_meta %>% left_join(map_table, by=c("seq_id" = "sample"))
(ps_kraken_species <- phyloseq(sample_data(s_meta),
                                otu_table(s_otu_tab, taxa_are_rows = TRUE),
                                tax_table(as.matrix(s_tax_tab))))

# filter taxa ("Remove taxa not seen more than 3 times in at least 20% of the samples" )
ps_kraken_species <-filter_taxa(ps_kraken_species, function(x) sum(x > 3) > (0.2*length(x)), TRUE)
ps_kraken_species
```

```{r}
library(metagenomeSeq)
heatmapCols = colorRampPalette(brewer.pal(9, "RdBu"))(50)

pdf("../../results/notebook_out/02A_KrakenCorrPlotGenera.pdf", width = 12, height=12)
plotCorr(obj = ps_kraken_species %>% phyloseq_to_metagenomeSeq(), n = 100, cexRow = 0.5, cexCol = 0.5,
trace = "none", dendrogram = "none", col = heatmapCols)
dev.off()


pdf("../../results/notebook_out/02A_KrakenAbundPlot.pdf")
plotMRheatmap(obj = ps_kraken_species %>% phyloseq_to_metagenomeSeq(), n = 200, cexRow = 0.25, cexCol = 0.25,
trace = "none", dendrogram = "none", col = heatmapCols)
dev.off()
```

```{r}
psmelt(ps_kraken_species) %>% 
  left_join(map_table, by=c("Sample"="sample")) %>% 
  filter(Genus=="Simplexvirus") %>%
  ggplot(., aes(x=sample.type, y=log(Abundance))) + geom_boxplot()
```
## MAASLIN2
```{r}
kraken_4mas <-psmelt(ps_kraken_species) %>% 
  select(Sample,Abundance, Genus) %>% 
  pivot_wider(names_from=Genus, values_from=Abundance) %>% 
  arrange(Sample) %>%
  column_to_rownames("Sample") %>% as.matrix()
meta <- read_table('../../config/map_table.tsv')

meta_4mas <-
  meta %>% 
  filter(!sample %in% control_samples) %>%
  column_to_rownames("sample")
meta_4mas$subject.id <- as.factor(meta_4mas$bal.number)
meta_4mas$sample.type <- as.factor(meta_4mas$sample.type)
meta_4mas$binned_outcome <- as.factor(meta_4mas$binned_outcome)

head(meta_4mas)
```

```{r}
fit_data2 = Maaslin2(input_data     = kraken_4mas, 
                     input_metadata = meta_4mas, 
                     min_abundance  = 5,
                     min_prevalence = .2,
                     normalization  = "CSS",
                     output         = "../../results/notebook_out/maaslin2_out", 
                     fixed_effects  = c("binned_outcome", "diagnosis.subtype"),
                     reference      = c("binned_outcome,Deceased", "diagnosis.subtype,NonPneumonia"),
                     random_effects = c("subject.id"),
                     core           = 4,
                     ) # Subject
```

# Consensus Cluster Analysis
```{r}

# Convert PS object (here not filtered) to metagenomeseq object
ms_kraken_species <- ps_kraken_species %>% phyloseq_to_metagenomeSeq()

# Normalize counts using CSS method
ms_kraken_norm<- MRcounts(ms_kraken_species, norm = TRUE, log = FALSE)

# Get top 500 most variable features using median absolute deviation method
mads=apply(ms_kraken_norm,1,mad)
d=ms_kraken_norm[rev(order(mads))[1:2000],]
d = sweep(d,1, apply(d,1,median,na.rm=T))
library(ConsensusClusterPlus)
report = ConsensusClusterPlus(d,maxK=15,reps=100,pItem=0.8,pFeature=1,clusterAlg="hc",distance="pearson",seed=123456)
report[[5]]
icl = calcICL(report)
icl[["itemConsensus"]]

# Cluster assignments for k=[[num]]
# Use enframe to convert from terrible list to tibble
clusters <- enframe(report[[8]]$consensusClass)
```

```{r}
# beta-diversity measure
beta <- vegdist(t(d), 'jaccard', binary = T)
beta.mat <- as.matrix(beta)

# projection
pcoa <- cmdscale(beta, k = 4, eig = TRUE)

# cleanup
ord <- as.data.frame(pcoa$points)
names(ord) <- c('pcoa1', 'pcoa2', 'pcoa3', 'pcoa4') ## rename coordinates

# add clusters

#group_by(item) %>% max(itemConsensus)

ord$Sample <- rownames(ord)

ord<-ord %>% left_join(clusters, by = c("Sample" = "name"))
# Percent explained variation
eig <- eigenvals(pcoa)
eig.percent <- 100*head(eig/sum(eig))
eig.percent
eig.1 <- paste("PCo1 (", as.character(round(eig.percent[1], digits = 1)), "%)", sep = "")
eig.2 <-paste("PCo2 (", as.character(round(eig.percent[2], digits = 1)), "%)", sep = "")
```

```{r}
## plot PCoA (FIGURE 1A)
ggplot(data = ord, aes(x = pcoa1, y = pcoa2, fill = as.factor(value))) +
  geom_point(size = 2, stroke = 1, shape=21, alpha=.5) +
  theme_bw() +
  xlab(eig.1) +
  ylab(eig.2) +
  theme_q2r() +
  theme(axis.text = element_text(size=15, color = "black"),
        axis.title = element_text(size=16, color = "black"),
        legend.text = element_text(size=15, color = "black"),
        legend.title = element_text(size=15, color = "black")) +
  scale_fill_brewer(guide="legend", palette="Set1") +
  #scale_shape_manual(values=c(21))+#, 22, 23, 24, 25, 26)) +
  guides(fill=guide_legend(override.aes=list(shape=21))) +
   theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) 
#ggsave("../../results/notebook_out/02A_KrakenPCoA.pdf", units="in", width=6, height=3.5)

```
```

```{r}
data(mouseData)
mouseData = mouseData[which(rowSums(mouseData)>139),]
sparsity = rowMeans(MRcounts(mouseData)==0)
lor = log(fitPA(mouseData,cl=pData(mouseData)[,3])$oddsRatio)
plotBubble(lor,sparsity,main="lor ~ sparsity")
# Example 2
x = runif(100000)
y = runif(100000)
plotBubble(y,x)
```


```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("ConsensusClusterPlus")
a```

