# SCRIPT OG Metagenomics Data

# 20230207

concatenated reads from Batch_04A1 and Batch_04A2

Starting analysis with updated processing

Making sample sheet
    python3 workflow/scripts/prep_sample_sheet.py --delimiter="_S" --path="/scratch/jsj3921/Batch_01" --subdirectories=True
    python3 workflow/scripts/prep_sample_sheet.py --delimiter="_S" --path="/scratch/jsj3921/Batch_02" --subdirectories=True --out="./config/sample_sheet_02.tsv"
    python3 workflow/scripts/prep_sample_sheet.py --delimiter="_S" --path="/scratch/jsj3921/Batch_03" --subdirectories=True --out="./config/sample_sheet_03.tsv"
    python3 workflow/scripts/prep_sample_sheet.py --delimiter="_R" --r1_common="R1" --r2_common="R2" --path="/scratch/jsj3921/Batch_04A12" --subdirectories=True --out="./config/sample_sheet_04.tsv"
    python3 workflow/scripts/prep_sample_sheet.py --delimiter="_R" --r1_common="R1" --r2_common="R2" --path="/scratch/jsj3921/Batch_04B" --subdirectories=True --out="./config/sample_sheet_04b.tsv"

Added 202221-Communal-Zymo to end of sample sheet as bioinformatic control

snakemake worked great

the following are 100% unknown from metaphlan 
     [1] "DNA_B03_21"  "DNA_B04_101" "DNA_B04_106" "DNA_B04_139" "DNA_B04_141" "DNA_B04_151" "DNA_B04_48"  "DNA_B04_61"  "DNA_B04_95"  "DNA_B04_98" 


# 20230109

Working on getting humann started
    Installing metaphlan 4 database required excessive RAM so had to do it in a batch job
    
    mamba activate /projects/b1180/software/conda_envs/humann/

    metaphlan 20221221-Comunal-Zymo.fastp_bowtie.r1.fastq.gz,20221221-Comunal-Zymo.fastp_bowtie.r2.fastq.gz --nproc 23 --input_type fastq -o mpatest.tsvls

Recent results from kraken2 and metaphlan output suggests a really interesting correlational relationship between groups of taxa 
This could be that there are taxa that have low competition with each other and thus co-associate and strong competition with others
that leads them to be co-exclusive
Some other taxa are just ok with being there independent of the other cohorts. Id hypothesize that these taxa provide some metabolic 
niche necessary for others and/or have different ecological constraints compared to the cohorts. That is, they can metabolize some 
rare but lung-abundant nutrient source. 
    Would be interesting if functional abundance profiles mimic the above, then find specific predictive metabolites

testing merging on humann/metaphlan output

module load BBMap
Usage for paired files:         bbmerge.sh in1=<read1> in2=<read2> out=<merged reads> outu1=<unmerged1> outu2=<unmerged2>
bbmerge.sh in1=20221221-Comunal-Zymo.fastp_bowtie.r1.fastq.gz in2=20221221-Comunal-Zymo.fastp_bowtie.r2.fastq.gz out=merged_test.fastq.gz outu1=unmergedout.1.fastq.gz outu2=unmergedout.2.fastq.gz

bbmerge.sh in1=20221221-Comunal-Zymo.fastp_bowtie.r1.fastq.gz \
    in2=20221221-Comunal-Zymo.fastp_bowtie.r2.fastq.gz \
    out=merged_test.fastq.gz \
    outu1=unmergedout.1.fastq.gz \
    outu2=unmergedout.2.fastq.gz \
    verystrict=t \
    k=60 \
    ihist=ihist.txt \
    mininsert=90
(will automatically make fast.gz)
- very strict = good for assembly based approached/reduced FP 
- k=60 is the kmer length used for overlap analysis/graph. Default is 31 but 60 is good for 150 PE
- mininsert=90 means that the min merged read length will be 90, which is noted in the literature as an optimal minimum for functional profiling

humann --input merged_test.fastq.gz --output humann_merged --threads 40 --memory-use maximum



bbmerge.sh in1=DNA_B04_115.fastp_bowtie.r1.fastq.gz \
    in2=DNA_B04_115.fastp_bowtie.r2.fastq.gz \
    out=merged_test.fastq.gz \
    outu1=unmergedout.1.fastq.gz \
    outu2=unmergedout.2.fastq.gz \
    verystrict=t \
    k=60 \
    ihist=ihist.txt \
    mininsert=90


# 20230210

Trying to add additional functionality to readQC module 

# 20230213

updated fastqc module to actually work

spades doesnt work with merged reads, so have to use PE reads

might try with megahit and use PE that dont merge in addition

with additional fastp parameters, complexity filter, etc additions to the pipeline, the n50
    decreased in spades outputs overall

# 20230214

Exploring freshly basecalled batch 4 samples to double check that nothign got corrupted along teh way

noticed that there is a wide distribution of reads in the samples, ranging from .fastq.gz files of size ~10M to ~1G

This may be the greater explanatory factor in why some samples just dont assemble well. 
    Must write code to visualize fastqc (and quast) multiqc outputs in order to contextualize/justify elimination

BCL Files in:
/scratch/jsj3921/bcl_files/SCRIPT_DNA_B04_PoolA_BCL_Files/RawFastq
/scratch/jsj3921/bcl_files/SCRIPT_DNA_B04_PoolA_Resequenced_BCL_Files/RawFastq
/scratch/jsj3921/bcl_files/SCRIPT_DNA_B04_PoolB_BCL_Files/RawFastq/RawFastq

PoolA1 starts with SCRIPT_DNA... so well keep that for now in order to differentiate between the sample types

Getting sample sheet ready
        python3 workflow/scripts/prep_sample_sheet.py --delimiter="_S" --path="/scratch/jsj3921/bcl_files/SCRIPT_DNA_B04_PoolA_BCL_Files/RawFastq"
        python3 workflow/scripts/prep_sample_sheet.py --delimiter="_S" --path="/scratch/jsj3921/bcl_files/SCRIPT_DNA_B04_PoolA_Resequenced_BCL_Files/RawFastq"
        python3 workflow/scripts/prep_sample_sheet.py --delimiter="_S" --path="/scratch/jsj3921/bcl_files/SCRIPT_DNA_B04_PoolB_BCL_Files/RawFastq/"


# 20230215

Got 16S data back!! Doing trim reads and then loading everything into qiime2

seq_processing env (for fastp)= /projects/b1042/HartmannLab/jack/mlm/.snakemake/conda/40fde712634bb6b71462835e30757718_

Ok so looking back at my script for mgx2amp it makes sense that we should just make a rule specific to amplicon
    so with that in mind, going to do the following:
    rule fastp4amp
    them copy to amp directory for analysis
    python3 workflow/scripts/prep_sample_sheet.py --delimiter="_S" --path="/projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq"

rule fastp4amp:
    """
    This is a special trimming rule used to quality trim and merge 
    reads from 16S amplicon sequencing. The remainder of my 16S pipeline 
    is not integrated into MLM (see MGX2AMP repo).

    After this step, prepare manifest and sample metadata files,
    acquire the SILVA DB for 16S sequences, and run QIIME2

    To use: 
        1. Comment out get_rules in rule all.
        2. Replace with expand(
            "results/AMP_trimmed/{sample}_fastp-merged.fq.gz", 
            sample=samples["sample"])
        3. Prep sample sheet and execute
    """
    input:
        r1 = get_r1,
        r2 = get_r2
    output:
        r1_merged = "results/AMP_trimmed/{sample}_fastp-merged.fq.gz",
        json = "results/AMP_trimmed/{sample}_fastp.json",
        html = "results/AMP_trimmed/{sample}_fastp.html"
    conda:
        "../envs/seq_processing.yml"
    threads: 15
    resources:
        mem="20G"
    shell: 
        """
        fastp \
            -i {input.r1} \
            -I {input.r2} \
            --merged_out {output.r1_merged} \
            --merge \
            --detect_adapter_for_pe \
            --trim_poly_g \
            --trim_poly_x \
            --thread {threads} \
            --length_required 250 \
            -j {output.json} \
            -h {output.html}
        """
successfully completed.
    Noticed some NTC samples have strickingly few reads after merging, etc.

DAMN IT I USED TRIMMED NOT MERGED READS FOR THE BETTER DADA2 PIPELINE

took out dedup
rule fastp_pe:
    input:
        r1 = get_r1,
        r2 = get_r2
    output:
        r1_filtered = "results/fastp_out/{sample}/{sample}.fastp.r1.fastq.gz",
        r2_filtered = "results/fastp_out/{sample}/{sample}.fastp.r2.fastq.gz",
        json = "results/fastp_out/{sample}/{sample}_fastp.json",
        html = "results/fastp_out/{sample}/{sample}_fastp.html"
    conda:
        "../envs/seq_processing.yml"
    threads: 12
    resources:
        mem="20G"
    shell: 
        """
        fastp \
            -i {input.r1} \
            -I {input.r2} \
            --out1 {output.r1_filtered} \
            --out2 {output.r2_filtered} \
            --detect_adapter_for_pe \
            --trim_poly_g \
            --trim_poly_x \
            --thread {threads} \
            --length_required 100 \
            -j {output.json} \
            -h {output.html} \
            -V 
        """

# 20230223

So the sample sheet that Hiams group gave me was wrong...indel "mutation" in sample order at plate2

So, I've updated the sample sheet and recalled/demultiplexed the samples again
    The old fastq folder was changed to RawFastq2Delete

We've regained lost water control Water_06 and have discovered 20230207NFW was not loaded 
    Instead, 2x 20230207Zymo controls were loaded (henceforth 20230207A_Zymo and 20230207B_Zymo)

Ran same sample sheet maker line as before. Differences below:
```
diff config/sample_sheet_v1amp.tsv config/sample_sheet.tsv

< 20230118_ZymoControl  MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230118_Zymo_Control_S92_R1_001.fastq.gz  /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230118_Zymo_Control_S92_R2_001.fastq.gz
< 20230201_Zymo2point5ng177     MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230201_Zymo_2point5ng_S177_R1_001.fastq.gz       /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230201_Zymo_2point5ng_S177_R2_001.fastq.gz
< 20230207_Zymo2point5ng272     MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230207_Zymo_2point5ng_S272_R1_001.fastq.gz       /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230207_Zymo_2point5ng_S272_R2_001.fastq.gz
< Water_230118NFW       MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/230118NFW_S93_R1_001.fastq.gz      /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/230118NFW_S93_R2_001.fastq.gz
< Water_230202NFW       MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/230202NFW_S178_R1_001.fastq.gz     /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/230202NFW_S178_R2_001.fastq.gz
< Water_230207NFW       MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/230207NFW_S273_R1_001.fastq.gz     /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/230207NFW_S273_R2_001.fastq.gz
< Water_F9Sham  MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/F9Sham_S163_R1_001.fastq.gz        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/F9Sham_S163_R2_001.fastq.gz
---
> 20230118_Zymo_Control MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230118_Zymo_Control_S92_R1_001.fastq.gz  /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230118_Zymo_Control_S92_R2_001.fastq.gz
> 20230201_Zymo_2point5ng       MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230201_Zymo_2point5ng_S177_R1_001.fastq.gz       /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230201_Zymo_2point5ng_S177_R2_001.fastq.gz
> 20230207A_Zymo_2point5ng      MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230207A_Zymo_2point5ng_S272_R1_001.fastq.gz      /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230207A_Zymo_2point5ng_S272_R2_001.fastq.gz
> 20230207B_Zymo_2point5ng      MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230207B_Zymo_2point5ng_S273_R1_001.fastq.gz      /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230207B_Zymo_2point5ng_S273_R2_001.fastq.gz
> 230118NFW     MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/230118NFW_S93_R1_001.fastq.gz      /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/230118NFW_S93_R2_001.fastq.gz
> 230202NFW     MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/230202NFW_S178_R1_001.fastq.gz     /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/230202NFW_S178_R2_001.fastq.gz
> 230207NFW     MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/230207NFW_S274_R1_001.fastq.gz     /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/230207NFW_S274_R2_001.fastq.gz
> F9Sham        MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/F9Sham_S163_R1_001.fastq.gz        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/F9Sham_S163_R2_001.fastq.gz
> Undetermined  MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/Undetermined_S0_R1_001.fastq.gz    /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/Undetermined_S0_R2_001.fastq.gz
```

- Delete Undetermined, Extra Underscore in Zymo
- Add 'Water_' to blanks and "272"/"273" to end of 07 zymo controls

Made updates, now differences are only based on S### value or A/B in case of zymo
    (As expected)
```
266c266,267
< 20230207_Zymo2point5ng272     MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230207_Zymo_2point5ng_S272_R1_001.fastq.gz       /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230207_Zymo_2point5ng_S272_R2_001.fastq.gz
---
> 20230207A_Zymo2point5ng272    MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230207A_Zymo_2point5ng_S272_R1_001.fastq.gz      /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230207A_Zymo_2point5ng_S272_R2_001.fastq.gz
> 20230207B_Zymo2point5ng273    MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230207B_Zymo_2point5ng_S273_R1_001.fastq.gz      /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/20230207B_Zymo_2point5ng_S273_R2_001.fastq.gz
269c270
< Water_230207NFW       MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/230207NFW_S273_R1_001.fastq.gz     /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/230207NFW_S273_R2_001.fastq.gz
---
> Water_230207NFW       MGX_DATA        /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/230207NFW_S274_R1_001.fastq.gz     /projects/b1042/HartmannLab/jack/balf_amplicon_microbiome/raw_fastq/RawFastq/230207NFW_S274_R2_001.fastq.g
```


