
nt Database	for kraken
mkdir kraken_db
module load aria/2 # multithreaded wget
 aria2c -x 10 https://genome-idx.s3.amazonaws.com/kraken/k2_nt_20230502.tar.gz
gunzip k2_nt_20230502.tar.gz 
tar -xvf k2_nt_20230502.tar 
# or instead of gunzip then tar use 
tar -xzvf k2_nt_20230502.tar.gz 

humann_join_tables -i results/humann_out -o merged_genefamilies.tsv --file_name genefamilies --search-subdirectories

humann_renorm_table --input merged_genefamilies.tsv --output merged_genefamilies-cpm.tsv --units cpm --update-snames
    params:
        sample_list = [i.split("/")[2] for i in {input.kraken_reports}]
 #[i.split("/")[2] for i in expand("results/kraken/{sample}/{sample}_kraken2report.txt", sample=samples["sample"])]


 The nt kraken2 db may actually be too big...
    needs like 650 g of RAM and take ~1hr to load database
    after loaded, very quick to execute sample processing
    Gave samples 1 hr and many of them failed to process


*** OBSERVATION: Poor assembly is perfectly correlated with spades warning: "None of paired reads aligned properly. Please, check orientation of your read pairs."
STD_B02_10ng-4534.out also has this warning
Batch 4 is most abundant in these warnings



sacctmgr show User jsj3921 --associations


# --- Dictionaries --- #
grep bowtie /projects/b1188/bmo/mlm/results/fastqc_out/multiqc_data/mqc_fastqc_sequence_duplication_levels_plot_1.txt 

multiqc --outdir results/fastqc_out --dirs --dirs-depth 2 results/fastqc_out/ -f

kneaddata --input1 DNA_B01_06_S6_L001_R1_001.fastq.gz --input2 DNA_B01_06_S6_L001_R2_001.fastq.gz --output ./kneadded --reference-db ../../resources/bowtie_human/chm13.draft_v1.0_plusY/ -t 20  --run-trim-repetitive --sequencer-source TruSeq3 --max-memory 50000 --fastqc /projects/b1188/bmo/mlm/results/kneaddata/kneaddata_env/bin/fastqc
mamba env create -f ../../workflow/envs/kneaddata.yml -p ./kneaddata_env

mamba install --force-reinstall java-jdk
  - java-jdk   8.0.92  1      installed                   
  + java-jdk   8.0.92  1      bioconda/linux-64     Cached

weird java error -- force reinsatlling fixed it...maybe have to add a line on recipe

bioconda::java-jdk=8.0.92
--output-prefix


kneaddata --input1 DNA_B01_06_S6_L001_R1_001.fastq.gz --input2 DNA_B01_06_S6_L001_R2_001.fastq.gz --output ./kneadded --reference-db ../../resources/bowtie_human/chm13.draft_v1.0_plusY/ -t 20 --run-trim-repetitive --sequencer-source TruSeq3 --fastqc /projects/b1188/bmo/mlm/results/kneaddata/kneaddata_env/bin/fastqc --output-prefix DNA_B01_06_keaddata --bowtie2-options "--very-sensitive"



See this link: github.com/conda/conda/issues/7980. These two lines of code at the beginning of your bash script will allow you to use conda activate in a script: CONDA_BASE=$(conda info --base) ; source $CONDA_BASE/etc/profile.d/conda.sh

ls metabat_spades_out/*/bins/*.bin.* -1 | grep -v "STD" | grep -v "unbinned" | grep -v "lowDepth" | grep -v "tooShort" | cut -d . -f 1 | uniq | wc -l

trying a newer megahit which is supposed to perform better 1.2.7
megahit -t 23 -m 0.9 -1 DNA_B01_06_keaddata_unmatched_1.fastq -2 DNA_B01_06_keaddata_unmatched_2.fastq --out-prefix DNA_B01_06 -o megahit_new
2023-09-06 14:31:29 - 8481 contigs, total 2691416 bp, min 269 bp, max 2334 bp, avg 317 bp, N50 310 bp
   11 contigs > 1kb

megahit -t 23 -m 0.9 -1 DNA_B01_06.bowtie.r1.fastq.gz -2 DNA_B01_06.bowtie.r2.fastq.gz --out-prefix DNA_B01_06 -o megahit_new
2023-09-06 14:50:39 - 5639 contigs, total 1708509 bp, min 269 bp, max 1217 bp, avg 302 bp, N50 296 bp
  2 contigs > 1kb

in local
    megahit -t 23 -m 0.9 -1 DNA_B01_06_S6_L001_R1_001_kneaddata_unmatched_1.fastq -2 DNA_B01_06_S6_L001_R1_001_kneaddata_unmatched_2.fastq --out-prefix DNA_B01_06 -o megahit_new
    2023-09-06 15:00:00 - 86 contigs, total 32464 bp, min 301 bp, max 861 bp, avg 377 bp, N50 360 bp


megahit -t 23 -m 0.9 -1 DNA_B01_03.bowtie.r1.fastq.gz -2 DNA_B01_03.bowtie.r2.fastq.gz --out-prefix DNA_B01_03 -o megahit_03
2023-09-06 17:27:39 - 35384 contigs, total 32122554 bp, min 205 bp, max 281661 bp, avg 907 bp, N50 1223 bp


kneaddata --input1 DNA_B01_03_S3_L001_R1_001.fastq.gz --input2 DNA_B01_06_S6_L001_R2_001.fastq.gz --output ./kneadded_03 --reference-db ../../resources/bowtie_human/chm13.draft_v1.0_plusY/ -t 23 --run-trim-repetitive --sequencer-source TruSeq3 --fastqc /projects/b1188/bmo/mlm/results/kneaddata/kneaddata_env/bin/fastqc --output-prefix DNA_B01_03 --bowtie2-options="--very-sensitive"


megahit -t 23 -m 0.9 -1 DNA_B01_03_unmatched_1.fastq -2 DNA_B01_03_unmatched_2.fastq --out-prefix DNA_B01_03 -o megahit_03

megahit -t 23 -m 0.9 -1 DNA_B01_03_unmatched_1.fastq -2 DNA_B01_03_unmatched_2.fastq --out-prefix DNA_B01_03 -o megahit_03
