# Notes 20230713

Working on several project, BMO, Hauser, and BAS/multiomics

**Goals**
- Add functional anlaysis using a gene catalog
    - IDEA: https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-022-01259-2
    - GET Gene catalog
        - RULE prokka annotation - get genes
        - RULES coassembly - get more genes (megahit)
        - RULE concat all genes (bash)
        - RULE cluster/deduplicate all genes (nonredundant, CDHit?, Orthogroup)
        - RULE functional annotation (eggNOG mapper)
    - GET functional table
        - RULE align samples to all genes (bowtie2)
        - RULE get feature counts (feature counts)
    - GET database QC
- Add functional analysis with HUMANN
- Fix metaphlan database setup issue
- ADD source tracker analysis
    - download stuff from sra using a list of accession number
    - setup source tracker in conda
    - run downloaded stuff through same mlm pipeline that is on 
    - use source tracker 
- REMOVE CONTAMINANTS
    

**Debugging**
- Working on snakemake 7.20.0. 
- Still having issue where snakemake cant activate conda env if using "script" directive. 
- Made clone of current snakemake env `mamba create --name snakemake_7.20.0 --clone snakemake`
- updating snakemake env to 7.30.1 (bug fixed in this version) `mamba install -n snakemake snakemake=7.30.1`
- fixed filter contigs script error. Solution was to add argparse functions and run with shell snakemake directive instead of script. 

**Prokka and Prodigal**
- Testing in interactive job `prokka scaffolds.fasta --cpus 23 --metagenome`
- Using module
- Version 1.14.6
- NOTE: prokka is wrapper of prodigal with function assignment integrated
- prokka needs smaller contigs names (<=37 bp) 
    awk line from chatgpt `awk '/^>/ { sub(/_length_[0-9]+_cov_[0-9.]+/, "", $0) } 1' scaffolds.fa`
- Prodigal `prodigal -i scaffolds.fasta -p meta -o sample.gbk`
    - prodigal works and gives best match in genesls
- NOTES on function
    - prokka is incldues fucntional annotation but has limited default databases
    - can specify that you want to include KEGG and GO 
    - headers only 37 length
    - change headers needed as step, can do in above awk
- Works! Prokka runs and have merge prokka outputs finalize
- COOL function found. Can subset which samples by a list of sample names in the cpnfig
    - Use this function to elim negative samples from being added to merged outputs
    - COULD ALSO USE TO ID NEGATIVE CONTROLS -> coassmeble -> remove contaminants
    -  expand("results/prokka_out/{sample}/{sample}.faa", sample=samples[~samples["sample"].isin(controls)]["sample"])
        - ~ is not
        - subset by vector-like object of str samples, `samples["sample"]`, for those in lst of strings `controls`
        - controls is in config
