
nt Database	for kraken
mkdir kraken_db
module load aria/2 # multithreaded wget
 aria2c -x 10 https://genome-idx.s3.amazonaws.com/kraken/k2_nt_20230502.tar.gz
gunzip k2_nt_20230502.tar.gz 
tar -xvf k2_nt_20230502.tar 
# or instead of gunzip then tar use 
tar -xzvf k2_nt_20230502.tar.gz 

humann_join_tables -i results/humann_out -o merged_genefamilies.tsv --file_name genefamilies --search-subdirectories

humann_renorm_table --input merged_genefamilies.tsv --output merged_genefamilies-cpm.tsv --units cpm --update-snames
    params:
        sample_list = [i.split("/")[2] for i in {input.kraken_reports}]
 #[i.split("/")[2] for i in expand("results/kraken/{sample}/{sample}_kraken2report.txt", sample=samples["sample"])]


 The nt kraken2 db may actually be too big...
    needs like 650 g of RAM and take ~1hr to load database
    after loaded, very quick to execute sample processing
    Gave samples 1 hr and many of them failed to process


*** OBSERVATION: Poor assembly is perfectly correlated with spades warning: "None of paired reads aligned properly. Please, check orientation of your read pairs."
STD_B02_10ng-4534.out also has this warning
Batch 4 is most abundant in these warnings



sacctmgr show User jsj3921 --associations


# --- Dictionaries --- #
grep bowtie /projects/b1188/bmo/mlm/results/fastqc_out/multiqc_data/mqc_fastqc_sequence_duplication_levels_plot_1.txt 

multiqc --outdir results/fastqc_out --dirs --dirs-depth 2 results/fastqc_out/ -f
